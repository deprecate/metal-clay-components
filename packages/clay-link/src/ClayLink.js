'use strict';

import Component from 'metal-component';
import 'clay-icon';
import Soy from 'metal-soy';
import { Config } from 'metal-state';

import templates from './ClayLink.soy.js';

/**
 * Implementation of the Metal Clay Link.
 */
class ClayLink extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayLink.STATE = {
	/**
	 * Sets the download attribute on the anchor tag.
	 * @instance
	 * @memberof ClayLink
	 * @type {?string|undefined}
	 * @default undefined
	 */
	download: Config.string(),

	/**
	 * Sets the HTML attribute href on the anchor tag.
	 * @instance
	 * @memberof ClayLink
	 * @type {?string|undefined}
	 * @default undefined
	 */
	href: Config.string(),

	/**
	 * Render MetalClayIcon in the link, available options are `alignment`,
	 * `elementClasses`, `spritemap`, `symbol`.
	 * @instance
	 * @memberof ClayLink
	 * @type {?Object|undefined}
	 * @default undefined
	 */
	icon: Config.object(),

	/**
	 * Id to be applied to the anchor tag.
	 * @instance
	 * @memberof ClayLink
	 * @type {?string|undefined}
	 * default undefined
	 */
	id: Config.string(),

	/**
	 * Sets the text or HTML to be rendered inside the anchor tag.
	 * @instance
	 * @memberof ClayLink
	 * @type {?string|undefined}
	 * @default undefined
	 */
	label: {
		isHtml: true,
	},

	/**
	 * Sets the HTML attribute target on the anchor tag.
	 * @instance
	 * @memberof ClayLink
	 * @type {?string|undefined}
	 * @default undefined
	 */
	target: Config.string(),
};

Soy.register(ClayLink, templates);

export { ClayLink };
export default ClayLink;
