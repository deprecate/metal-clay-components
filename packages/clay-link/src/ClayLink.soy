{namespace ClayLink}

/**
 * This renders an anchor element.
 */
{template .render}
	{@param? ariaLabel: string}
	{@param? buttonStyle: string}
	{@param? download: string}
	{@param? href: string}
	{@param? icon: [
		alignment: string,
		monospaced: bool,
		spritemap: string,
		symbol: string
	]}
	{@param? id: string}
	{@param? label: html|string}
	{@param? style: string}
	{@param? target: string}

	{let $attributes kind="attributes"}
		{if $ariaLabel}
			aria-label="{$ariaLabel}"
		{/if}

		{if $buttonStyle}
			class="btn btn-{$buttonStyle}"
		{elseif $style}
			class="link-{$style}"
		{/if}

		{if $download}
			download="{$download}"
		{/if}

		{if $href}
			href="{$href}"
		{/if}

		{if $id}
			id="{$id}"
		{/if}

		{if $target}
			target="{$target}"
		{/if}
	{/let}

	<a {$attributes}>
		{let $iconAlignment: $icon and $icon.alignment ? $icon.alignment : 'left' /}

		{if $icon and $iconAlignment == 'left'}
			{call .icon }
				{param icon: $icon /}
			{/call}
		{/if}

		{$label}

		{if $icon and $iconAlignment == 'right'}
			{call .icon }
				{param icon: $icon /}
			{/call}
		{/if}
	</a>
{/template}

/**
 * This renders a Clay Icon.
 */
{template .icon}
	{@param? icon: [
		alignment: string,
		monospaced: bool,
		spritemap: string,
		symbol: string
	]}

	{call ClayIcon.render}
		{param monospaced: $icon.monospaced /}
		{param spritemap: $icon.spritemap /}
		{param symbol: $icon.symbol /}
	{/call}
{/template}