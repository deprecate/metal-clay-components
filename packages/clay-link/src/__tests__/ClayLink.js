'use strict';

import dom from 'metal-dom';
import ClayLink from '../ClayLink';

let component;

describe('ClayLink', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should apply classes to the anchor tag', () => {
		component = new ClayLink({
			elementClasses: 'my-test-class',
		});

		expect(dom.hasClass(component.element, 'my-test-class')).toBeTruthy();
	});

	it('should apply the download attribute to the anchor tag', () => {
		component = new ClayLink({
			download: 'http://www.example.com',
		});

		expect(component.element.getAttribute('download')).toEqual(
			'http://www.example.com',
		);
	});

	it('should apply an id to the anchor tag', () => {
		component = new ClayLink({
			id: 'myUniqueId',
		});

		expect(component.element.getAttribute('id')).toEqual('myUniqueId');
	});

	it('should apply an href to the anchor tag', () => {
		component = new ClayLink({
			href: 'http://www.example.com',
		});

		expect(component.element.getAttribute('href')).toEqual(
			'http://www.example.com',
		);
	});

	it('should apply the target attribute to the anchor tag', () => {
		component = new ClayLink({
			target: 'http://www.example.com',
		});

		expect(component.element.getAttribute('target')).toEqual(
			'http://www.example.com',
		);
	});

	it('should render the correct markup with icon on the right', () => {
		component = new ClayLink({
			href: '#1',
			icon: {
				alignment: 'right',
				spritemap:
					'../node_modules/lexicon-ux/build/images/icons/icons.svg',
				symbol: 'add-cell',
			},
			label: '<span>My Other Link Label</span>',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render the correct markup with icon on the left', () => {
		component = new ClayLink({
			href: '#1',
			icon: {
				alignment: 'left',
				spritemap:
					'../node_modules/lexicon-ux/build/images/icons/icons.svg',
				symbol: 'add-cell',
			},
			label: 'My Link Label',
		});

		expect(component).toMatchSnapshot();
	});
});
