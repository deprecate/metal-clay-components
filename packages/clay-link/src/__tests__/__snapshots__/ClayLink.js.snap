// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayLink should render the correct markup with icon on the left 1`] = `
<a href="#1">
  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-add-cell">
    <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#add-cell"></use>
  </svg>My Link Label</a>
`;

exports[`ClayLink should render the correct markup with icon on the right 1`] = `
<a href="#1">
  <span>My Other Link Label</span>
  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-add-cell">
    <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#add-cell"></use>
  </svg>
</a>
`;
